### DISCUSSION TITLE
Splash screen of Java application stuck with "assign [class=...]"

### DISCUSSION THREAD

**Discussion by andrei-korshikov** [i3/i3#5985] [Q&A] [answered]
I have default i3 config with one line added:

    assign  [class="^FreeMind$"]  number  $ws3

FreeMind is a Java application, which is working with Java 11 only.

When $ws3 is active, everything is OK. Splash screen appears, splash screen disappears, main application window appears.

When $ws3 is not active (not shown on the screen) the strange thing happens. When I go to $ws3, I see **two** windows: one with splash screen (can't be closed even with $mod+Shift+q), and the second is main application window.

I had very little Java experience years ago, so maybe you have ideas or suggestions on what is going on?.

Just in case, both splash screen and main window have the same `class`, and both don't have `window_role`, so it's not possible to distinguish between them.

The program is open sourced, so I think I'll end up with disabling splash screen. But maybe there is better solution?

Edit: Maybe somehow related: https://github.com/i3/i3/discussions/5020

**Comment by waterkip** [comment]
Not the answer: I would look into freeplane, which is a fork of freemind, but it is maintained. Freemind has been 10 years without maintenance, whereas free plane is receiving updates.

**Reply by andrei-korshikov** [reply]
(Offtopic) In principle, I do agree, and would definitely recommend newcomers to start with FreePlane.

From my perception (of course, it is extremely subjective) "FreeMind vs (another tool)" is like "i3 vs Desktop Environment".

I wouldn't say FreeMind has been without maintenance, it's just rock solid and needed almost no maintenance at all; when there was a major problem with Java 17 - is has been resolved quickly (in unstable). New features are rejected on purpose.

And, frankly, I even do not use all the features of FreeMind itself - no colors, no icons, no tables - all this stuff distracts me (yes, perception is subjective, I know, just describing my experience), black text on white background - that's all what I need. And I have been using FreeMind for over 15 years on mostly daily basis, of course I'm too addicted:D

**Comment by waterkip** [comment]
Can you check of the splash screen has a title? For example, with the Garmin SDK I do things like this:

```
for_window [class="Simulator" title="Connect IQ Device Simulator"] floating enabled
```

You could probably do something like this too?

**Reply by waterkip** [reply]
I tested it (with Freeplane). The splash screen title is win0. I didn't have weird things on my workspace even if I set the splash screen to $media (which wasn't the workspace where I started the app).

I think just making the splash screen floating would cause fewer issues. Then, assign it to a workspace and "Bob's your uncle". 

```
for_window [class="org-knopflerfish-framework-BundleThread" title="win0" ] floating enable
assign [class="org-knopflerfish-framework-BundleThread" title="win0"] workspace $cli
assign [class="org-knopflerfish-framework-BundleThread" ] workspace $media
```

**Reply by andrei-korshikov** [reply, marked as answer]
Make splash screen floating is nice, but it still does not solve the original issue: if the target workspace is not shown — then splash screen becomes noncloseable nice floating window.

By the way, I've tested with the latest source (November 2023) and the latest Java (22), so it's definitely not a Java bug, but an issue of the FreeMind itself (yes, Freeplane does not have this problem).

Finally I have a workaround that satisfies me — a tiny shell script:
```shell
#!/bin/sh
i3-msg 'workspace "3: Mind Map "'
/usr/bin/freemind $1
```

The idea is pretty simple — switch to the workspace at first, then start the application.

In theory, it should be possible in one line, something like `i3-msg 'workspace ...; exec freemind'` — like in the User's Guide, but that works only _sometimes_: it's like exec is started too early, at the moment the workspace is not created. I.e. at the end the target workspace is created, but empty, and the application is started in the current workspace.

And, of course, I totally appreciate making the splash screen floating:
```
for_window [class="^FreeMind$" title="^FreeMind$"] floating enable
```

**Reply by waterkip** [reply]
You could try it with one line with `exec --no-startup-id`, eg, `i3-msg "workspace 3; exec --no-startup-id /path/to/freemind"`

**Reply by andrei-korshikov** [reply]
Hmm… I've seen no difference. Why do you think `--no-startup-id` does something special to that case? I'm still thinking that's about "exec is starting too early", but of course it can just depend on my environment…

**Reply by andrei-korshikov** [reply]
By the way, I thought that Java applications are not "startup-notification" aware. Maybe I'm wrong and that thought was from really old days…

(Freeplane offtopic) I've tried it, and had a really epic fail:( My X server crashes when I try to change the zoom with Control + mouse wheel.
Also, when I open my FreeMind files in Freeplane, fonts look at least strange or even ugly. I use mind maps as a my knowledge base (like wiki, but in mind map format), so… I can edit and convert all of them, but it's quite a big amount of work.
So, I'll give Freeplane another try in another few years:))

**Reply by waterkip** [reply]
The `--no-startup-id` will start the application and drop it somewhere regardless of where you executed the command (in the workspace or elsewhere). Meaning, if you have something similar to `assign [class="yourfreemindclass"] workspace $mindmapping`, regardless of where you start the application, it will go that that workspace.

I had some issues with and/without `--no-startup-id`. I use a tool that applies a layout to a workspace if it hasn't been opened yet and starts up application that are swallowed by the layout. Without the `--no-startup-id` I've had some issues, which is why I thought that might help you. I guess it didn't.

Read about it here: https://i3wm.org/docs/userguide.html#exec

**Reply by andrei-korshikov** [reply]
Very interesting finding (I mean the difference with with/without)! I hope I'll remember it. Anyway, thank you.
